File: src/main/java/miragefairy2019/colormaker/core/PanelColorSlider.java
package miragefairy2019.colormaker.core;

import kotlin.Unit;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ComponentAdapter;
import java.awt.event.ComponentEvent;
import java.awt.image.BufferedImage;
import java.util.ArrayList;
import java.util.function.Consumer;
import java.util.regex.Pattern;

import static mirrg.boron.swing.UtilsComponent.get;

public class PanelColorSlider extends JPanel {

    private PanelSliderField sliderG;
    private PanelSliderField sliderB;
    private PanelSliderField sliderR;
    private ParsingTextField<Integer> textField;

    private boolean isInProcessing = false;

    public PanelColorSlider() {

        setLayout(get(new GridBagLayout(), l -> {
            l.columnWidths = new int[]{
                    0, 0
            };
            l.rowHeights = new int[]{
                    0, 0, 0, 0
            };
            l.columnWeights = new double[]{
                    1.0, Double.MIN_VALUE
            };
            l.rowWeights = new double[]{
                    0.0, 0.0, 0.0, 0.0
            };
        }));

        add(get(sliderR = new PanelSliderField(), c -> {
            c.listeners.add(value -> {
                if (isInProcessing) return;
                setValue(new Color(sliderR.getValue(), sliderG.getValue(), sliderB.getValue()), c);
            });
        }), get(new GridBagConstraints(), c -> {
            c.insets = new Insets(0, 0, 5, 0);
            c.fill = GridBagConstraints.BOTH;
            c.gridx = 0;
            c.gridy = 0;
            c.gridwidth = 2;
            c.gridheight = 1;
        }));

        add(get(sliderG = new PanelSliderField(), c -> {
            c.listeners.add(value -> {
                if (isInProcessing) return;
                setValue(new Color(sliderR.getValue(), sliderG.getValue(), sliderB.getValue()), c);
            });
        }), get(new GridBagConstraints(), c -> {
            c.insets = new Insets(0, 0, 5, 0);
            c.fill = GridBagConstraints.BOTH;
            c.gridx = 0;
            c.gridy = 1;
            c.gridwidth = 2;
            c.gridheight = 1;
        }));

        add(get(sliderB = new PanelSliderField(), c -> {
            c.listeners.add(value -> {
                if (isInProcessing) return;
                setValue(new Color(sliderR.getValue(), sliderG.getValue(), sliderB.getValue()), c);
            });
        }), get(new GridBagConstraints(), c -> {
            c.insets = new Insets(0, 0, 5, 0);
            c.fill = GridBagConstraints.BOTH;
            c.gridx = 0;
            c.gridy = 2;
            c.gridwidth = 2;
            c.gridheight = 1;
        }));

        Pattern pattern = Pattern.compile("[0-9a-fA-F]{6}");
        add(get(textField = new ParsingTextField<>(s -> {
            if (pattern.matcher(s.trim()).matches()) {
                return Integer.parseInt(s.trim(), 16);
            } else {
                return null;
            }
        }, v -> String.format("%06X", v & 0xffffff)), c -> {
            c.setColumns(10);
            c.getListeners().add(i -> {
                if (isInProcessing) return Unit.INSTANCE;
                setValue(new Color(c.getValue()), c);
                return Unit.INSTANCE;
            });
        }), get(new GridBagConstraints(), c -> {
            c.fill = GridBagConstraints.HORIZONTAL;
            c.gridx = 0;
            c.gridy = 3;
            c.gridwidth = 1;
            c.gridheight = 1;
        }));

        add(get(new JToggleButton("Pick"), new Consumer<JToggleButton>() {
            private Timer timer = new Timer(20, e -> {
                try {
                    Point location = MouseInfo.getPointerInfo().getLocation();
                    BufferedImage createScreenCapture = new Robot().createScreenCapture(new Rectangle(location.x, location.y, 1, 1));
                    setValue(new Color(createScreenCapture.getRGB(0, 0)));
                } catch (Exception e2) {
                    e2.printStackTrace();
                }
            });

            {
                timer.setRepeats(true);
            }

            @Override
            public void accept(JToggleButton c) {
                c.addActionListener(e -> {
                    if (c.isSelected()) {
                        timer.start();
                    } else {
                        timer.stop();
                    }
                });
                c.addComponentListener(new ComponentAdapter() {
                    @Override
                    public void componentHidden(ComponentEvent e) {
                        timer.stop();
                    }
                });
            }
        }), get(new GridBagConstraints(), c -> {
            c.fill = GridBagConstraints.HORIZONTAL;
            c.gridx = 1;
            c.gridy = 3;
            c.gridwidth = 1;
            c.gridheight = 1;
        }));

        //

        setValue(Color.white);

    }

    //

    private Color value;

    public void setValue(Color value) {
        setValue(value, null);
    }

    public Color getValue() {
        return value;
    }

    //

    public ArrayList<Consumer<Color>> listeners = new ArrayList<>();

    private void setValue(Color value, Object source) {
        isInProcessing = true;

        this.value = value;
        if (source != textField) textField.setValue(value.getRGB());
        if (source != sliderR) sliderR.setValue(value.getRed());
        if (source != sliderG) sliderG.setValue(value.getGreen());
        if (source != sliderB) sliderB.setValue(value.getBlue());
        listeners.forEach(l -> {
            try {
                l.accept(value);
            } catch (Exception e) {
                e.printStackTrace();
            }
        });

        isInProcessing = false;
    }

}

File: src/main/java/miragefairy2019/colormaker/core/PanelColorSlider.kt
package miragefairy2019.colormaker.core;

import kotlin.Unit;

import javax.swing.JPanel;
import javax.swing.JToggleButton;
import javax.swing.Timer;
import java.awt.Color;
import java.awt.GridBagConstraints;
import java.awt.GridBagLayout;
import java.awt.Insets;
import java.awt.MouseInfo;
import java.awt.Point;
import java.awt.Rectangle;
import java.awt.Robot;
import java.awt.event.ComponentAdapter;
import java.awt.event.ComponentEvent;
import java.awt.image.BufferedImage;
import java.util.ArrayList;
import java.util.function.Consumer;
import java.util.regex.Pattern;

import mirrg.boron.swing.UtilsComponent.get;

class PanelColorSlider : JPanel {

    private lateinit var sliderG: PanelSliderField;
    private lateinit var sliderB: PanelSliderField;
    private lateinit var sliderR: PanelSliderField;
    private lateinit var textField: ParsingTextField<Int>;

    private var isInProcessing = false;

    constructor() {

        setLayout(get(GridBagLayout(), { l ->
            l.columnWidths = intArrayOf(
                    0, 0
            );
            l.rowHeights = intArrayOf(
                    0, 0, 0, 0
            );
            l.columnWeights = doubleArrayOf(
                    1.0, Double.MIN_VALUE
            );
            l.rowWeights = doubleArrayOf(
                    0.0, 0.0, 0.0, 0.0
            );
        }));

        add(get(PanelSliderField().also { sliderR = it }, { c ->
            c.listeners.add(java.util.function.IntConsumer { value ->
                if (isInProcessing) return@IntConsumer;
                setValue(Color(sliderR.getValue(), sliderG.getValue(), sliderB.getValue()), c);
            });
        }), get(GridBagConstraints(), { c ->
            c.insets = Insets(0, 0, 5, 0);
            c.fill = GridBagConstraints.BOTH;
            c.gridx = 0;
            c.gridy = 0;
            c.gridwidth = 2;
            c.gridheight = 1;
        }));

        add(get(PanelSliderField().also { sliderG = it }, { c ->
            c.listeners.add(java.util.function.IntConsumer { value ->
                if (isInProcessing) return@IntConsumer;
                setValue(Color(sliderR.getValue(), sliderG.getValue(), sliderB.getValue()), c);
            });
        }), get(GridBagConstraints(), { c ->
            c.insets = Insets(0, 0, 5, 0);
            c.fill = GridBagConstraints.BOTH;
            c.gridx = 0;
            c.gridy = 1;
            c.gridwidth = 2;
            c.gridheight = 1;
        }));

        add(get(PanelSliderField().also { sliderB = it }, { c ->
            c.listeners.add(java.util.function.IntConsumer { value ->
                if (isInProcessing) return@IntConsumer;
                setValue(Color(sliderR.getValue(), sliderG.getValue(), sliderB.getValue()), c);
            });
        }), get(GridBagConstraints(), { c ->
            c.insets = Insets(0, 0, 5, 0);
            c.fill = GridBagConstraints.BOTH;
            c.gridx = 0;
            c.gridy = 2;
            c.gridwidth = 2;
            c.gridheight = 1;
        }));

        val pattern = Pattern.compile("[0-9a-fA-F]{6}");
        add(get(ParsingTextField({ s ->
            if (pattern.matcher(s.trim()).matches()) {
                return@ParsingTextField Integer.parseInt(s.trim(), 16);
            } else {
                return@ParsingTextField null;
            }
        }, { v -> String.format("%06X", v and 0xffffff) }).also { textField = it }, { c ->
            c.setColumns(10);
            c.listeners.add({ i ->
                if (isInProcessing) return@add Unit;
                setValue(Color(c.value!!), c);
                return@add Unit;
            });
        }), get(GridBagConstraints(), { c ->
            c.fill = GridBagConstraints.HORIZONTAL;
            c.gridx = 0;
            c.gridy = 3;
            c.gridwidth = 1;
            c.gridheight = 1;
        }));

        add(get(JToggleButton("Pick"), object : Consumer<JToggleButton> {
            private val timer = Timer(20, { e ->
                try {
                    val location = MouseInfo.getPointerInfo().getLocation();
                    val createScreenCapture = Robot().createScreenCapture(Rectangle(location.x, location.y, 1, 1));
                    setValue(Color(createScreenCapture.getRGB(0, 0)));
                } catch (e2: Exception) {
                    e2.printStackTrace();
                }
            });

            init {
                timer.setRepeats(true);
            }

            override fun accept(c: JToggleButton) {
                c.addActionListener({ e ->
                    if (c.isSelected()) {
                        timer.start();
                    } else {
                        timer.stop();
                    }
                });
                c.addComponentListener(object : ComponentAdapter() {
                    override fun componentHidden(e: ComponentEvent) {
                        timer.stop();
                    }
                });
            }
        }), get(GridBagConstraints(), { c ->
            c.fill = GridBagConstraints.HORIZONTAL;
            c.gridx = 1;
            c.gridy = 3;
            c.gridwidth = 1;
            c.gridheight = 1;
        }));

        //

        setValue(Color.white);

    }

    //

    private lateinit var value: Color;

    fun setValue(value: Color) {
        setValue(value, null);
    }

    fun getValue(): Color {
        return value;
    }

    //

    @JvmField val listeners : ArrayList<Consumer<Color>> = ArrayList();

    private fun setValue(value: Color, source: Any?) {
        isInProcessing = true;

        this.value = value;
        if (source != textField) textField.setValue(value.getRGB());
        if (source != sliderR) sliderR.setValue(value.getRed());
        if (source != sliderG) sliderG.setValue(value.getGreen());
        if (source != sliderB) sliderB.setValue(value.getBlue());
        listeners.forEach({ l ->
            try {
                l.accept(value);
            } catch (e: Exception) {
                e.printStackTrace();
            }
        });

        isInProcessing = false;
    }

}
